### PR TITLE
Mesh 1570/clean old scope_id

### PR DESCRIPTION
**Pull request by vkomenda**
### modified logic

- Calling the `add_investor_uniqueness_claim` extrinsic in the `identity` pallet with a new investor uniqueness `claim` that has a new `ScopeId` cleans up identity-ticker balances related to the `ScopeId` of the old investor uniqueness claim if there was one.

### CODE REVIEW

File: pallets/identity/src/lib.rs [hunk 1]
```diff
@@ -1651,28 +1646,48 @@ impl<T: Trait> Module<T> {
...
         ensure!(
             ValidProofOfInvestor::evaluate_claim(&claim, &target, &proof),
             Error::<T>::InvalidScopeClaim
         );
 
-        if let Claim::InvestorUniqueness(Scope::Ticker(scope), scope_id, _) = &claim {
+        if let Scope::Ticker(ticker) = &scope {
+            // If `target` already has different InvestorUniqueness claims, clean up the old ScopeId data.
+            let old_scope_ids = Self::fetch_base_claims(target, ClaimType::InvestorUniqueness)
+                .filter_map(move |id_claim| {
```
Comment (Centril) [inline pallets/identity/src/lib.rs:1674]: This is fine, but I'd probably write this using `match` + a guard myself in the current language.

File: pallets/runtime/tests/src/identity_test.rs [hunk 2]
```diff
@@ -1557,3 +1558,69 @@ fn add_permission_with_secondary_key() {
...
+    let no_balance_at_scope = |scope_id| {
+        assert_eq!(
+            false,
+            <pallet_asset::BalanceOfAtScope<TestStorage>>::contains_key(scope_id, alice.did)
+        );
+    };
+    let balance_at_scope = |scope_id, balance| {
+        assert_eq!(
+            balance,
+            <pallet_asset::BalanceOfAtScope<TestStorage>>::get(scope_id, alice.did)
```
Comment (Centril) [inline pallets/runtime/tests/src/identity_test.rs:1586]:
```suggestion
            Asset::balance_of_at_scope(scope_id, alice.did)
```

File: pallets/runtime/tests/src/identity_test.rs [hunk 3]
```diff
@@ -1557,3 +1558,69 @@ fn add_permission_with_secondary_key() {
...
+        );
+    };
+    let scope_id_of = |scope_id| {
+        assert_eq!(scope_id, <pallet_asset::ScopeIdOf>::get(ticker, alice.did));
+    };
+    let aggregate_balance = |scope_id, balance| {
+        assert_eq!(
+            balance,
+            <pallet_asset::AggregateBalance<TestStorage>>::get(ticker, scope_id)
```
Comment (Centril) [inline pallets/runtime/tests/src/identity_test.rs:1595]:
```suggestion
            Asset::aggregate_balance(ticker, scope_id)
```

File: pallets/runtime/tests/src/identity_test.rs [hunk 4]
```diff
@@ -1557,3 +1558,69 @@ fn add_permission_with_secondary_key() {
...
+        );
+    };
+    let balance_at_scope = |scope_id, balance| {
+        assert_eq!(
+            balance,
+            <pallet_asset::BalanceOfAtScope<TestStorage>>::get(scope_id, alice.did)
+        );
+    };
+    let scope_id_of = |scope_id| {
+        assert_eq!(scope_id, <pallet_asset::ScopeIdOf>::get(ticker, alice.did));
```
Comment (Centril) [inline pallets/runtime/tests/src/identity_test.rs:1590]:
```suggestion
        assert_eq!(scope_id, Asset::scope_id_of(ticker, alice.did));
```

File: pallets/runtime/tests/src/identity_test.rs [hunk 5]
```diff
@@ -1557,3 +1558,69 @@ fn add_permission_with_secondary_key() {
...
+    ExtBuilder::default()
+        .cdd_providers(vec![AccountKeyring::Charlie.public()])
+        .build()
+        .execute_with(|| do_add_investor_uniqueness_claim());
+}
+
+fn do_add_investor_uniqueness_claim() {
+    let alice = User::new(AccountKeyring::Alice);
+    let cdd_provider = AccountKeyring::Charlie.public();
+    let ticker = simple_asset(alice.acc(), true).1;
```
Comment (Centril) [inline pallets/runtime/tests/src/identity_test.rs:1573]: Let's avoid adding more technical debt and use `basic_asset` instead.

File: pallets/identity/src/lib.rs [hunk 6]
```diff
@@ -1651,28 +1646,48 @@ impl<T: Trait> Module<T> {
...
         // Verify the confidential claim.
         ensure!(
             ValidProofOfInvestor::evaluate_claim(&claim, &target, &proof),
             Error::<T>::InvalidScopeClaim
         );
 
-        if let Claim::InvestorUniqueness(Scope::Ticker(scope), scope_id, _) = &claim {
+        if let Scope::Ticker(ticker) = &scope {
+            // If `target` already has different InvestorUniqueness claims, clean up the old ScopeId data.
+            let old_scope_ids = Self::fetch_base_claims(target, ClaimType::InvestorUniqueness)
```
Comment (maxsam4) [inline pallets/identity/src/lib.rs:1673]: Instead of having this unbounded iterator and filtering here, you should directly query `asset.ScopeIdOf`.

**Review by maxsam4 — CHANGES_REQUESTED** [review]
LGTM overall but need a change that will simplify a few things.

File: pallets/runtime/tests/src/identity_test.rs (hunk 5 above)
Comment (vkomenda) [reply pallets/runtime/tests/src/identity_test.rs:1573]: I don't get your point. To use `basic_asset` I would need to define a `SecurityToken` anyway which is what `simple_asset` is doing.

Comment (Centril) [reply pallets/runtime/tests/src/identity_test.rs:1573]: `fn simple_asset` duplicates code and shouldn't exist. You're adding another dependency to it. Instead, use `a_token` to produce `(Ticker, SecurityToken<u128>)` to pass to `basic_asset`, or if you prefer, create `fn asset(owner: User) -> Ticker` as a wrapper for that logic.

Comment (vkomenda) [reply pallets/runtime/tests/src/identity_test.rs]: Done. Also marked `simple_asset` as deprecated.

**Review by maxsam4 — APPROVED** [review]

File: pallets/identity/src/lib.rs [hunk 7]
```diff
@@ -1651,28 +1646,31 @@ impl<T: Trait> Module<T> {
         proof: InvestorZKProofData,
         expiry: Option<T::Moment>,
     ) -> DispatchResult {
+        let (scope, scope_id, cdd_id) =
+            if let Claim::InvestorUniqueness(scope, scope_id, cdd_id) = &claim {
+                (scope, scope_id, cdd_id)
+            } else {
+                return Err(Error::<T>::ClaimVariantNotAllowed.into());
```
Comment (fmiguelgarcia) [inline pallets/identity/src/lib.rs:1653]:
```suggestion
                fail!(Error::<T>::ClaimVariantNotAllowed);
```

**Review by fmiguelgarcia — APPROVED** [review]

**Review by Centril — APPROVED** [review]

**Pull request merged**